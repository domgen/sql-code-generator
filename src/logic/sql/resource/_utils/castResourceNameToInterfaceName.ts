import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../../model';

export const castResourceNameToInterfaceName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
