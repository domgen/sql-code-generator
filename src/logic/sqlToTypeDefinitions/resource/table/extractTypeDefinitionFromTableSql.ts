import strip from 'sql-strip-comments';

import { castCommasInParensToPipesForTokenSafety } from '../common/castCommasInParensToPipesForTokenSafety';
import { extractTypeDefinitionFromColumnSql } from './extractTypeDefinitionFromColumnSql';
import { TypeDefinitionOfResourceTable } from '../../../../model';

/*
  note:
    - for table and view, just a map of { property: type }
    - for function and procedure, input type, output type
*/

const MYSQL_KEY_OR_CONSTRAINT_LINE_REGEX = /[KEY|key].*\s\(/g;

export const extractTypeDefinitionFromTableSql = ({ name, sql }: { name: string; sql: string }) => {
  // 0. strip comments
  const strippedSql: string = strip(sql);

  // 1. grab the insides of the "create" (i.e., 'CREATE TABLE ... (__INSIDES__) ...' => '__INSIDES__')
  const innerSqlAndAfter = strippedSql
    .split('(')
    .slice(1)
    .join('('); // drop the part before the first '('
  const innerSql = innerSqlAndAfter
    .split(')')
    .slice(0, -1)
    .join(')'); // drop the part after the last ')'

  // 2. cast commas inside of parens into pipes, so that we treat them as unique tokens when splitting "property lines" by comma
  const parenTokenizedInnerSql = castCommasInParensToPipesForTokenSafety({ sql: innerSql });

  // 3. grab definition lines, by splitting out properties by commas
  const lines = parenTokenizedInnerSql.split(',');

  // 4. strip out the lines that are defining keys or constraints
  const columnDefinitions = lines
    .filter((line) => {
      return !new RegExp(MYSQL_KEY_OR_CONSTRAINT_LINE_REGEX, 'gm').test(line); // note, we reinstantiate so as to not share regex object (i.e., state) between calls
    })
    .map((line) => line.trim());

  // 5. get column definition from each property
  const columns = columnDefinitions.map((line) => extractTypeDefinitionFromColumnSql({ sql: line }));

  // 6. return the table definition
  return new TypeDefinitionOfResourceTable({
    name,
    columns,
  });
};
