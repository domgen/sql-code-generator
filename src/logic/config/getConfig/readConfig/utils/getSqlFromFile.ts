import { readFileAsync } from '../../../_utils/readFileAsync';

export const getSqlFromFile = async ({ filePath }: { filePath: string }): Promise<string> => {
  const extension = filePath.split('.').slice(-1)[0];
  if (extension === 'sql') return readFileAsync({ filePath });
  if (extension === 'ts') {
    const exports = await import(filePath);
    if (!exports.sql) throw new Error('.ts file did not have an export with name "sql"');
    return exports.sql;
  }
  throw new Error('path must lead to file with extension of .ts or .sql');
};
