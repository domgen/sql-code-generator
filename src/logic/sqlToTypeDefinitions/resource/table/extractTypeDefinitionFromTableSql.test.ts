import { extractTypeDefinitionFromTableSql } from './extractTypeDefinitionFromTableSql';
import { getSqlFromFile } from '../../../config/_utils/getSqlFromFile';

describe('extractTypeDefinitionFromTableSql', () => {
  it('should be able to extract types in this example', async () => {
    const exampleSql = await getSqlFromFile({ filePath: `${__dirname}/../../../__test_assets__/tables/image.sql` });
    const typeDef = extractTypeDefinitionFromTableSql({ name: 'image', sql: exampleSql });
    expect(typeDef).toMatchSnapshot();
  });
  it('should be able to extract types in this other example', async () => {
    const exampleSql = await getSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/tables/suggestion_version.sql`,
    });
    const typeDef = extractTypeDefinitionFromTableSql({ name: 'suggestion_version', sql: exampleSql });
    expect(typeDef).toMatchSnapshot();
  });
});
