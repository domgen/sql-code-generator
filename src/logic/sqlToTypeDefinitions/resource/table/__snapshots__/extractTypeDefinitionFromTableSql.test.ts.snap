// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTypeDefinitionFromTableSql mysql should be able to extract types in this example 1`] = `
TypeDefinitionOfResourceTable {
  "columns": [
    TypeDefinitionOfResourceColumn {
      "name": "id",
      "type": [
        "number",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "uuid",
      "type": [
        "string",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "created_at",
      "type": [
        "Date",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "url",
      "type": [
        "string",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "caption",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "credit",
      "type": [
        "string",
        "null",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "alt_text",
      "type": [
        "string",
        "null",
      ],
    },
  ],
  "name": "image",
}
`;

exports[`extractTypeDefinitionFromTableSql mysql should be able to extract types in this other example 1`] = `
TypeDefinitionOfResourceTable {
  "columns": [
    TypeDefinitionOfResourceColumn {
      "name": "id",
      "type": [
        "number",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "suggestion_id",
      "type": [
        "number",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "effective_at",
      "type": [
        "Date",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "created_at",
      "type": [
        "Date",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "status",
      "type": [
        "string",
      ],
    },
  ],
  "name": "suggestion_version",
}
`;

exports[`extractTypeDefinitionFromTableSql postgres should be able to extract types in this example 1`] = `
TypeDefinitionOfResourceTable {
  "columns": [
    TypeDefinitionOfResourceColumn {
      "name": "id",
      "type": [
        "number",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "uuid",
      "type": [
        "string",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "created_at",
      "type": [
        "Date",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "url",
      "type": [
        "string",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "description",
      "type": [
        "string",
        "null",
      ],
    },
  ],
  "name": "photo",
}
`;

exports[`extractTypeDefinitionFromTableSql postgres should be able to extract types in this other example 1`] = `
TypeDefinitionOfResourceTable {
  "columns": [
    TypeDefinitionOfResourceColumn {
      "name": "id",
      "type": [
        "number",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "job_id",
      "type": [
        "number",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "effective_at",
      "type": [
        "Date",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "created_at",
      "type": [
        "Date",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "location_id",
      "type": [
        "number",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "photo_ids_hash",
      "type": [
        "Buffer",
      ],
    },
  ],
  "name": "job_version",
}
`;

exports[`extractTypeDefinitionFromTableSql postgres should be able to extract types in this other example, having a check constraint with nested parens 1`] = `
TypeDefinitionOfResourceTable {
  "columns": [
    TypeDefinitionOfResourceColumn {
      "name": "id",
      "type": [
        "number",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "uuid",
      "type": [
        "string",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "created_at",
      "type": [
        "Date",
      ],
    },
    TypeDefinitionOfResourceColumn {
      "name": "status",
      "type": [
        "string",
      ],
    },
  ],
  "name": "job",
}
`;
